Fixed cmake to make dependencies optional

DiamonDinoia opened the pull request:
Hello,

I do not think nda should force the user to have these dependencies if we want to make it more general as a project.
I _hacked_ my way out to make it compile without. I think this is useful in general to any project that just needs a tensor library and they can optin for additional dependencies.

Cheers,
Marco

DiamonDinoia committed:
Fixed cmake to make dependencies optional

DiamonDinoia committed:
fixes

DiamonDinoia committed:
fixed cmake, parallelism dispatch, more tests

DiamonDinoia committed:
removing asan from combination

DiamonDinoia committed:
fixed typo

DiamonDinoia commented:
@parcollet could you have a look at this? The failing test might be a bug in MPI

parcollet commented on c++/nda/sym_grp.hpp:
```diff
@@ -176,24 +180,64 @@ namespace nda {
-        // distribute data among all ranks
+        a() = 0.0;
+        for (auto const &sym_class : mpi::chunk(sym_classes)) {
+          auto idx           = a.indexmap().to_idx(sym_class[0].first);
```
@Thoemi09 Can you have a look at this with Marco ? It seems the mpi::chunk is causing an error ?

Thoemi09 replied:
I think the `#pragma omp parallel for` loops are the problem.

@DiamonDinoia Can you either try using a `#pragma omp parallel` instead or removing the `itertools::omp_chunk` calls? The second option is probably cleaner.

DiamonDinoia replied:
Hi @Thoemi09,

I am not sure what you mean here:
I made both mpi and openmp optional.  The previous loop was:
```cpp
#pragma omp parallel
        for (auto const &sym_class : itertools::omp_chunk(mpi::chunk(sym_classes))) {
```
is now split into 4 depending on the parallelism enabled:
BOTH MPI and OMP, same as before:
```cpp
#pragma omp parallel for
        for (auto const &sym_class : itertools::omp_chunk(mpi::chunk(sym_classes))) {
```
MPI only:

```cpp
// no pragma 
      for (auto const &sym_class : mpi::chunk(sym_classes)) {
```
OpenMP only:
```cpp
#pragma omp parallel for
        for (auto const &sym_class : itertools::omp_chunk(sym_classes)) {
```

Could you clarify?

Thoemi09 replied:
Hey @DiamonDinoia,

You changed `#pragma omp parallel` to `#pragma omp parallel for`. For this to work, you also have to remove the `itertools::omp_chunk` calls.

`omp_chunk` just splits a range into subranges depending on the OMP thread ID that calls it. If it is used in addition to`#pragma omp parallel for`, I would assume that the loop is only done over one (the first) of the subranges.

The test failures indicate that just a part of the matrix/tensor is initialized correctly, e.g.
```
   Value of: array_are_close(A, B)
    Actual: false (max_element(abs(X - Y)) = 0.96886777112423139
   X = 
  [[(0.1354770042967805,0),(0.2210340429827049,0),(0.2210340429827049,0),(0.1354770042967805,0)]
   [(0.8350085899945795,0),(0.30816705050700327,0),(0.30816705050700327,0),(0.8350085899945795,0)]
   [(0.1354770042967805,0),(0.2210340429827049,0),(0.2210340429827049,0),(0.1354770042967805,0)]
   [(0.96886777112423139,0),(0.54722059636785192,0),(0.54722059636785192,0),(0.96886777112423139,0)]]
   Y = 
  [[(0.1354770042967805,0),(0,0),(0,0),(0.1354770042967805,0)]
   [(0,0),(0.30816705050700327,0),(0.30816705050700327,0),(0,0)]
   [(0.1354770042967805,0),(0,0),(0,0),(0.1354770042967805,0)]
   [(0,0),(0,0),(0,0),(0,0)]])
  Expected: true
```
I think that's because we are missing some of the symmetry classes from the for-loop.

DiamonDinoia replied:
AHA! Well spotted. I misunderstood the API. Let me test this.

DiamonDinoia committed:
fixed omp issue & merged unstable

DiamonDinoia committed:
fixed CI

DiamonDinoia committed:
fixed CI, missing newline

DiamonDinoia committed:
Finding h5 twice breaks macos

DiamonDinoia commented:
@parcollet, @Wentzell or @Thoemi09  This is ready for review. Could any of you have a look and merge?

DiamonDinoia committed:
remove extra tab

Thoemi09 committed:
Enable HDF5, MPI and OpenMP by default

Thoemi09 committed:
Minor changes to CMakeLists.txt files

- minimize changes and try to keep them consistent

Thoemi09 committed:
Undo some include changes

- keep the ./ for local includes to be consistent with all other includes
- add back missing include to mpi/reduce.hpp

Thoemi09 committed:
Rename MPI_SUPPORT to NDA_HAVE_MPI

- in analogy to NDA_HAVE_CUDA
- add NDA_HAVE_OPENMP and NDA_HAVE_H5

Thoemi09 committed:
Simplify changes in sym_grp.hpp

- keep runtime parallelization flag and use whatever is available

Thoemi09 committed:
Add tests for sym_grp with MPI support

DiamonDinoia committed:
changing defaults

DiamonDinoia commented:
@Thoemi09 I changed the defaults to be enabled if the dependencies are found on the system. Otherwise warn the user and go on. I think this might be the best for everyone.

Wentzell committed:
Iterate on refactor-cmake PR, various changes

Wentzell committed:
[ghactions] Set SDKROOT environment variable

Wentzell committed:
Run nda_sym_grp test also in mpi modeif possible

Wentzell reviewed (approved):
Thank you @DiamonDinoia @Thoemi09 for these improvements! I have iterated on the PR and everything looks good now
